nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/util/FlowDifferenceFilters.java

```diff
@@ -547,4 +577,302 @@ private static boolean isParameterContextChange(final FlowDifference flowDiffere
     private static boolean isLogFileSuffixChange(final FlowDifference flowDifference) {
         return flowDifference.getDifferenceType() == DifferenceType.LOG_FILE_SUFFIX_CHANGED;
     }
+
+    public static EnvironmentalChangeContext buildEnvironmentalChangeContext(final Collection<FlowDifference> differences, final FlowManager flowManager) {
+        if (differences == null || differences.isEmpty() || flowManager == null) {
+            return EnvironmentalChangeContext.empty();
+        }
+
+        final Set<String> serviceIdsReferencedByNewProperties = new HashSet<>();
+        final Map<String, List<PropertyDiffInfo>> parameterizedAddsByComponent = new HashMap<>();
+        final Map<String, List<PropertyDiffInfo>> parameterizationRemovalsByComponent = new HashMap<>();
+
+        for (final FlowDifference difference : differences) {
+            if (difference.getDifferenceType() == DifferenceType.PROPERTY_ADDED) {
+                final ComponentNode componentNode = Optional.ofNullable(getComponent(flowManager, difference.getComponentB()))
+                        .orElseGet(() -> getComponent(flowManager, difference.getComponentA()));
+                if (componentNode != null) {
+                    final Optional<String> fieldNameOptional = difference.getFieldName();
+                    if (fieldNameOptional.isPresent()) {
+                        final PropertyDescriptor propertyDescriptor = componentNode.getPropertyDescriptor(fieldNameOptional.get());
+                        if (propertyDescriptor != null && !propertyDescriptor.isDynamic() && propertyDescriptor.getControllerServiceDefinition() != null) {
+                            final Object valueB = difference.getValueB();
+                            if (valueB instanceof String) {
+                                serviceIdsReferencedByNewProperties.add((String) valueB);
```

Comment:
Minor adjustment to remove the need for casting:
```suggestion
                            if (valueB instanceof String serviceId) {
                                serviceIdsReferencedByNewProperties.add(serviceId);
```